crates/top/re_sdk/src/lenses/builder.rs

```diff
@@ -0,0 +1,335 @@
+        let output_builder =
+            ScatterColumnsBuilder::new(ast::TargetEntity::Explicit(entity_path.into()));
+        let output = builder(output_builder).build(&self.input)?;
+        self.outputs.push(output);
+        Ok(self)
+    }
+
+    /// Finalizes this builder and returns the corresponding lens.
+    pub fn build(self) -> ast::Lens {
+        ast::Lens {
+            input: self.input,
+            outputs: self.outputs,
+        }
+    }
+}
+
+// ==================== Output Builders ====================
+
+/// Builder for temporal outputs with 1:1 row mapping.
+///
+/// Each input row produces exactly one output row. Outputs inherit time columns
+/// from the input, plus any additional time columns specified.
+pub struct ColumnsBuilder {
+    target_entity: ast::TargetEntity,
+    components: Vec<ast::ComponentOutput>,
+    time_outputs: Vec<ast::TimeOutput>,
+}
+
+impl ColumnsBuilder {
+    fn new(target_entity: ast::TargetEntity) -> Self {
+        Self {
+            target_entity,
+            components: vec![],
+            time_outputs: vec![],
+        }
+    }
+
+    /// Adds a component output column.
+    ///
+    /// # Arguments
+    /// * `component_descr` - The descriptor for the output component
+    /// * `ops` - Sequence of operations to apply to transform the input column
+    pub fn component(
+        mut self,
+        component_descr: ComponentDescriptor,
+        ops: impl IntoIterator<Item = Op>,
+    ) -> Self {
+        self.components.push(ast::ComponentOutput {
+            component_descr,
+            ops: ops.into_iter().collect(),
+        });
+        self
+    }
+
+    /// Adds a time extraction.
+    ///
+    /// Extracts data from the input column to create a new time column for the output rows.
+    ///
+    /// # Arguments
+    /// * `timeline_name` - Name of the timeline to create
+    /// * `timeline_type` - Type of timeline (Sequence or Time)
+    /// * `ops` - Sequence of operations to extract time values (must produce [`arrow::array::Int64Array`])
+    pub fn time(
+        mut self,
+        timeline_name: impl Into<TimelineName>,
+        timeline_type: TimeType,
+        ops: impl IntoIterator<Item = Op>,
+    ) -> Self {
+        self.time_outputs.push(ast::TimeOutput {
+            timeline_name: timeline_name.into(),
+            timeline_type,
+            ops: ops.into_iter().collect(),
+        });
+        self
+    }
+
+    /// Builds a [`LensKind`], the `input` is passed for providing contextualized errors.
+    fn build(self, input: &ast::InputColumn) -> Result<ast::LensKind, LensError> {
+        Ok(ast::LensKind::Columns(OneToOne {
+            target_entity: self.target_entity,
+            components: self.components.try_into().map_err(|_err| {
+                LensError::MissingOutputComponent {
+                    // TODO(grtlr): This should not use debug formatting.
+                    input_filter: format!("{:?}", input.entity_path_filter),
```

Comment:
Why use formatting at all here?

```suggestion
                    input_filter: input.entity_path_filter.clone(),
```